### PR TITLE
Add guided encrypted partitioning option.

### PR DESCRIPTION
**Pull request by xnox**
https://asciinema.org/a/uOeYHMVkFmBtfl7c9YLUJ2LPm

Not sure about where to store the guided passphrase, hence it's tagged onto the controller. Note that it is cleared any time one goes back. As ultimately it's the property of the volume group inside the model.

Also not sure about the text on the "password" screen and the help text on the choose disk screen.

### CODE REVIEW

**Comment by xnox** [comment]
ascinema thing has old wording, and maybe i should stay on the new pages for longer....

Stealing wording from ubiquity.

**Comment by mwhudson** [comment]
Thanks for working on this!

This makes me think of a dusty old bug about LVM not using the whole disk (https://bugs.launchpad.net/subiquity/+bug/1785321). The proposed fix there is to ask the user how much of the disk to put in the initial VG. The reason for bringing this up is to ask: should that be asked on the same screen as the passphrase for the encrypted case? Should there even just be one screen that asks how much of the disk to use, passphrase and which disk to use and keys whether to use encryption off a passphrase is entered? Just rambling really but interested to see what you think.

(This all reminds me that the guided options would be more useful if we resurrected my old hairy code to be more flexible about allowing objects to be edited even if they are used...)

**Comment by mwhudson** [comment]
Code looks fine btw, modulo some slightly strange wrapping.

File: subiquity/ui/views/filesystem/guided.py [hunk 1]
```diff
@@ -98,13 +111,72 @@ def cancel(self, btn=None):
...
+
+    def validate_confirm_passwor(self):
+        if self.password.value != self.confirm_password.value:
+            return _("Security keys do not match")
+
+class GuidedPasswordView(BaseView):
+    title = _("Choose a security key")
+    footer = _("")
+    excerpt = _("""Disk encryption protects your files in case you lose your \
+computer. It requires you to enter a security key each time the \
```
Comment (mwhudson) [inline subiquity/ui/views/filesystem/guided.py]: Wrapping seems odd here.

Comment (xnox) [reply subiquity/ui/views/filesystem/guided.py]:
```
>>> print("""long string \
... without leading chars.
... 
... and again long string \
...      but with leading indentation.
... """)
long string without leading chars.

and again long string      but with leading indentation.

```

Comment (xnox) [reply subiquity/ui/views/filesystem/guided.py]: i'm not sure how to wrap things, without introducing additional whitespace, whilst also wrapping to a reasonable source code line width.

**Comment by xnox** [comment]
> Thanks for working on this!
> 
> This makes me think of a dusty old bug about LVM not using the whole disk (https://bugs.launchpad.net/subiquity/+bug/1785321). The proposed fix there is to ask the user how much of the disk to put in the initial VG. The reason for bringing this up is to ask: should that be asked on the same screen as the passphrase for the encrypted case? Should there even just be one screen that asks how much of the disk to use, passphrase and which disk to use and keys whether to use encryption off a passphrase is entered? Just rambling really but interested to see what you think.
> 
> (This all reminds me that the guided options would be more useful if we resurrected my old hairy code to be more flexible about allowing objects to be edited even if they are used...)

At the very least changing the default from 4GB to 80% would be a massive improvement.
I'm not sure if we should be asking this, or if it should be amendable on the next screen of manual partitioning / confirmation. Ie. you can change the size of the '/' lvm on the next screen.

**Comment by vorlonofportland** [comment]
> At the very least changing the default from 4GB to 80% would be a massive improvement.

Agreed.  Sensible default is more important than configurability of the LV size during install; and the flow should give you the option of not thinking about the LV size during install, even if you make it configurable.  And 100% is NOT a sensible default, because if you're using LVM it's because you want to have space for snapshots or additional LVs and you can't do an online shrink of a filesystem so an LV using 100% is useless.  But on any modern disk, 80% means you have space that you could lift-and-shift the rootfs to a new LV to shrink it if you decide post-install that you want something different.

**Comment by mwhudson** [comment]
OK, so I think the plan there should be to have a better default (and maybe make changing it afterwards easier). So then for this branch do we ask for the passphrase on the same screen as the disk selection? I guess probbbbbably not.

**Comment by xnox** [comment]
> OK, so I think the plan there should be to have a better default (and maybe make changing it afterwards easier). So then for this branch do we ask for the passphrase on the same screen as the disk selection? I guess probbbbbably not.

So, I think we need design input for this. I have mimicked what is done in Ubiquity. They have one screen to pick the guide options (full disk install, dual-boot install, manual, etc), with tick boxes to "and use LVM" or like "and encrypt", with separate pages to do disk selection, and separate page for the password.

Merging the password and disk selection pages might be doable, but it might make the screen scrollable due to help text for the password and the disk selection......

**Comment by xnox** [comment]
Things to fix.
* create form wrapper for radio button group
* use it
* restructure dialogs

I have started doing some of things things locally already, but not yet fully done.

**Comment by xnox** [comment]
Started doing the new UI as per mpt request.... https://github.com/CanonicalLtd/subiquity/pull/588 and it's a bit hairy.

I wonder, if we can land rebased this code as is, (method+password+disk selection in three steps) and the proceed to rework the guided UI later (method+disk selection in a single step).

**Comment by mwhudson** [comment]
This is all done in separate branches now.

**Pull request closed without merging**